archive/imo/imo2014_q1.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2021 Adrián Doña Mateo. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adrián Doña Mateo
+-/
+
+import algebra.big_operators
+import data.pnat.basic
+
+/-!
+# IMO 2014 Q1
+
+Let a₀ < a₁ < a₂ < ⋯ be an infinite sequence of positive integers.
+Prove that there exists a unique n ≥ 1 such that
+
+  aₙ < (a₀ + a₁ + ⋯ + aₙ) / n < aₙ₊₁.
+
+This solutions is a translation of the official solution, which may be found as the
+solution to problem A1 [here](https://www.imo-official.org/problems/IMO2014SL.pdf).
+-/
+
+open_locale big_operators
+
+noncomputable theory
+
+/- The integer sequence. -/
+variable (a : ℕ → ℤ)
+
+/- We define an auxiliary sequence `d a : ℕ+ → ℤ` by
+  `d a n = (a 0 + a 1 + ⋯ + a n) - n * (a n)`. -/
+def d (n : ℕ+) : ℤ := (∑ i : fin (n + 1), a i) - n * (a n)
+
+lemma first_ineq_iff {n : ℕ+} :
+  (a n : ℚ) < (↑∑ i : fin (n + 1), a i : ℚ) / n ↔ 0 < d a n :=
+begin
+  have : (↑(a n) : ℚ) = ↑(a n) / ↑(1 : ℤ) := by simp,
+  change _ < _ / ↑(n : ℤ) ↔ _,
+  rw [this, rat.div_lt_div_iff_mul_lt_mul], swap,
+  { norm_num }, swap,
+  { norm_cast, have : 1 ≤ ↑n := pnat.one_le n, linarith },
+  simp [d], rw mul_comm
+end
+
+lemma second_ineq_iff {n : ℕ+} :
+  (↑∑ i : fin (n + 1), a i : ℚ) / n ≤ a (n + 1) ↔ d a (n + 1) ≤ 0 :=
+begin
+  have : (↑(a (n + 1)) : ℚ) = ↑(a (n + 1)) / ↑(1 : ℤ) := by simp,
+  change _ / ↑(n : ℤ) ≤ _ ↔ _, rw this,
+  simp [rat.div_num_denom],
+  rw rat.le_def, swap,
+  { norm_cast, have : 1 ≤ ↑n := pnat.one_le n, linarith }, swap,
+  { norm_num },
+  simp [d, @fin.sum_univ_cast_succ _ _ ↑(n + 1)],
+  rw [add_mul, one_mul],
+  simp,	rw mul_comm, refl,
+end
+
+/- We rephrase the original question into a question about `d a`. -/
+lemma ineq_iff {n : ℕ+} :
+  (a n : ℚ) < (↑∑ i : fin (n + 1), a i : ℚ) / n ∧
+  (↑∑ i : fin (n + 1), a i : ℚ) / n ≤ a (n + 1)
+  ↔ 0 < d a n ∧ d a (n + 1) ≤ 0 :=
+⟨λ h, ⟨(first_ineq_iff a).1 h.1, (second_ineq_iff a).1 h.2⟩,
+  λ h, ⟨(first_ineq_iff a).2 h.1, (second_ineq_iff a).2 h.2⟩⟩
+
+lemma d_one : d a 1 = a 0 :=
+show a 0 + (a 1 + 0) - 1 * (a 1) = a 0, by ring
+
+/- If `a` is strictly increasing, then `d a` is strictly decreasing. -/
+lemma ddes (hinc : ∀ n, a n < a (n + 1)) : ∀ n, d a (n + 1) < d a n :=
+begin
+  intro n,
+  have : ↑n * (a n - a (n + 1)) < 0 :=
+    mul_neg_iff.mpr (or.inl ⟨by simp, sub_lt_zero.mpr (hinc n)⟩),
+  have hneg : d a (n + 1) - d a n < 0, from
+    calc ((∑ i : fin (n + 2), a i) - (n + 1) * (a (n + 1)))
+          - ((∑ i : fin (n + 1), a i) - n * (a n))
+          = ((∑ i : fin (n + 1), a i) + a (n + 1) - (n + 1) * (a (n + 1)))
+          - ((∑ i : fin (n + 1), a i) - n * (a n))
+      : by simp [fin.sum_univ_cast_succ]
+    ... = a (n + 1) - (n + 1) * (a (n + 1)) + n * (a n)                  : by ring
+    ... = n * (a n - a (n + 1))                                          : by ring
+    ... < 0                                                              : this,
```

Comment:
I think there are no precise rules here, but it seems more common in mathlib to have a bit less space between the equation and the `:  by`.